test/client.go

```diff
@@ -0,0 +1,86 @@
+package main
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	rpc "github.com/keybase/go-framed-msgpack-rpc"
+	"golang.org/x/net/context"
+)
+
+type GenericClient interface {
+	Call(ctx context.Context, method string, arg interface{}, res interface{}) error
+	Notify(ctx context.Context, method string, arg interface{}) error
+}
+
+//---------------------------------------------------------------------
+
+type TestClient struct {
+	GenericClient
+}
+
+func (a TestClient) Add(ctx context.Context, arg AddArgs) (ret int, err error) {
+	err = a.Call(ctx, "test.1.testp.add", arg, &ret)
+	return
+}
+
+func (a TestClient) Broken() (err error) {
+	err = a.Call(nil, "test.1.testp.broken", nil, nil)
+	return
+}
+
+func (a TestClient) UpdateConstants(ctx context.Context, arg Constants) (err error) {
+	err = a.Notify(ctx, "test.1.testp.updateConstants", arg)
+	return
+}
+
+func (a TestClient) GetConstants(ctx context.Context) (ret Constants, err error) {
+	err = a.Call(ctx, "test.1.testp.GetConstants", nil, &ret)
+	return
+}
+
+func (a TestClient) LongCall(ctx context.Context) (ret int, err error) {
+	err = a.Call(ctx, "test.1.testp.LongCall", nil, &ret)
+	return
+}
+
+//---------------------------------------------------------------------
+
+type Client struct {
+	port int
+}
+
+func (s *Client) Run() (err error) {
+	var c net.Conn
+	if c, err = net.Dial("tcp", fmt.Sprintf("127.0.0.1:%d", s.port)); err != nil {
+		return
+	}
+
+	xp := rpc.NewTransport(c, nil, nil)
+	cli := TestClient{GenericClient: rpc.NewClient(xp, nil)}
+
+	for A := 10; A < 23; A += 2 {
+		var res int
+		if res, err = cli.Add(context.Background(), AddArgs{A: A, B: 34}); err != nil {
+			return
+		}
+		fmt.Printf("result is -> %v\n", res)
+	}
+
+	err = cli.Broken()
+	fmt.Printf("for broken: %v\n", err)
+
+	if err = cli.UpdateConstants(context.Background(), Constants{Pi: 314}); err != nil {
+		return err
+	}
+	time.Sleep(3 * time.Millisecond)
```

Comment:
Replying to your comment:

> You can't RPC channels (at least it's meaningless to do so), and since Notify responds immediately, there's no guarantee that the UpdateConstants call will have completed when GetConstants is called.

Right, so I didn't mean to RPC over a channel. I mean something like wrap TestClient's GenericClient field into a struct that writes to an unbuffered channel on every RPC. So the test code would look like this:

```
blockingClient := BlockingClient{rpc.NewClient(xp. nil)}
cli := TestClient{GenericClient: blockingClient}
cli.UpdateConstants(...)
<-blockingClient.onRPC
// Here we know UpdateConstants has finished
cli.GetConstants(...)
```